### PR TITLE
first draft at correctly fixing #442

### PR DESCRIPTION
**Pull request by ancientHacker**
So actually unreconciling the off-currency transacation is easy, but I believe the containing document should be marked dirty when we do this and I don't see an obvious, clean way to do that.  @hsoft can you give me some advice?

### CODE REVIEW

**Comment by hsoft** [comment]
I was rather thinking of doing this only at load time (which is the only moment that this situation can happen since the rest of the app is locked-down in this matter). I'd see this code in https://github.com/hsoft/moneyguru/blob/master/core/loader/base.py#L230 instead.

It seems to me that with this PR, we're checking for this problem needlessly often (cooking happens all the time).

**Comment by ancientHacker** [comment]
I think I disagree, for the following reasons:

- We are talking about a file-format invariant here (no reconciliation of off-currency transactions).  The right place to monitor and maintain file-format invariants is *always*, because bugs that introduce them can introduce them at any time.
- Given how long it took to close off all the ways a user can do these off-currency reconciliations, it would not be wise to assume that there aren't any left.  (Plus, future bugs could re-enable them.)  This fix won't help in those cases.
- As a matter of principle, I believe that one should never do a sanity check at load time that one doesn't do at save time.  Better to do it in one place that maintains the invariant than to do the same check in two places (one on input, one on output).
- There is no actual performance hit doing this in cook.  In fact, the same check is done by every currency addition (hence the ValueErrors that were being thrown).  If you are worried about this check being needless in most cases (since it's done by the addition), I am happy to rewrite it as a catch of the ValueError that gets thrown.  But I think it's in the right place: where the file-format invariant matters.

Also, regardless of which way we do it, we still need to mark the resulting document dirty (with set_dirty, I assume).  I agree that doing this check on load makes it easier to do that, but I'm guessing that it's pretty easy to find the current open document when doing the cooking as well?

**Comment by hsoft** [comment]
Oh, this is getting interesting! Let's disagree then :) We'll have to number our points though.

1. True, it's possible that a new or existing bug make us violate a rule that we've established, that is "no reconciliation of off-currency transactions".
2. True, but moneyGuru has a very good automated test suite. New features are usually well covered test-wise and regressions and *always* tested.
3. Not a bad idea, I'll develop later.
4. You're probably right, but performance is not my main concern here, it's rather the *idea* of needless processing.

Now, me:

1. My proposition to do this check as load-time was under the assumption that moneyGuru is pretty well locked-up in this matter and that a bug making us violate our rule is unlikely.
2. But if we accept that this is a false assumption, then this present solution irks me because it would silently ignore the bug that causes the rule violation in the first place.
3. So, what we could do is to do your check as load time, unreconcile silently (the cause is likely to be that the file was saved with an old version of moneyGuru) *and* at save time, this time by popping up a dialog explaining that we had to unreconcile some transactions because something went wrong and that this problem should be reported.
4. If we do that, there's no need to ever touch the dirty flag.

What do you think?

**Comment by ancientHacker** [comment]
Yay, this is getting fun!  I think we are substantially aligned; let me try to outline something I believe we can both agree on (using numbers again :)...

1. This is an "older version corruption" problem that we should look for and fix when we load every file.
2. We should _not_ mark a fixed file dirty when we load it, because dirtiness implies that closing without save and then reloading the original will produce different data, and that's not true in this case.
3. We don't _need_ to check for off-currency reconciliations at save, because any such introduced by future bugs will get caught by the failure in cook.
4. We don't _want_ to warn about off-currency reconciliations at save, because it's incredibly scary for users to get warnings at Save time about possible bugs in the program they are using.  In my opinion, that's actually the _worst_ time to warn them.
4. We _must not_ fix off-currency reconciliations at save, because to change user-entered data at the time it's saved breaks the cardinal rule of what Save means.

If we are concerned about the far-fetched possibility that future bugs will re-introduce off-currency reconciliations that somehow don't get caught in cook before the file is saved, closed, and re-loaded from disk, then the right way to do that is to introduce a version tag into the file format (if there isn't one already; all I see is the document ID but maybe that encodes the version somehow) and have the load-time fix check the version tag and warn the user if the version tag is above 2.9 and off-currency reconciliations were found.  But I really think that's overkill! :)

In summary, I think we can just move my fix to load time and be done with it.  That OK with you?

**Comment by hsoft** [comment]
Yes, we agree :) Also, I've only received that crash report once, so the versioning thing is probably overkill, but it sounds like an elegant idea to the problem in case we are confronted to it again later.
